File: example/android/app/src/main/java/com/taobao/power_image_example/GlideMultiFrameImage.java
package com.taobao.power_image_example;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.drawable.Drawable;

import com.bumptech.glide.load.resource.gif.GifDrawable;
import com.taobao.power_image.loader.FlutterMultiFrameImage;

import java.lang.reflect.Field;
import java.lang.reflect.Method;

/**
 */
public class GlideMultiFrameImage extends FlutterMultiFrameImage {


    private static Field sGifState;
    private static Method sCurrentFrameMethod;
    private static Field sFrameloader;

    static {
        try {
            sGifState = GifDrawable.class.getDeclaredField("state");
            sGifState.setAccessible(true);
            sFrameloader = sGifState.getType().getDeclaredField("frameLoader");
            sFrameloader.setAccessible(true);
            sCurrentFrameMethod = sFrameloader.getType().getDeclaredMethod("getCurrentFrame");
            sCurrentFrameMethod.setAccessible(true);
        } catch (Throwable  e1) {
            e1.printStackTrace();
        }
    }

    private final Canvas tmpCanvas = new Canvas();


    public GlideMultiFrameImage(GifDrawable drawable, boolean needRecycle) {
        super(drawable,  needRecycle);

    }

    @Override
    public Bitmap getCurrentFrame(Drawable who) {
        who.draw(tmpCanvas);
        Bitmap bitmap = null;
        try {
            Object gifState = sGifState.get(who);
            Object frameLoader = sFrameloader.get(gifState);
            bitmap = (Bitmap) sCurrentFrameMethod.invoke(frameLoader);
        } catch (Throwable t) {
            t.printStackTrace();
        }

        return bitmap;
    }

    @Override
    protected void onStart(Drawable who) {
        ((GifDrawable) who).start();
    }

    @Override
    protected void onRelease(Drawable who) {
        ((GifDrawable) who).stop();
    }

    @Override
    public int getFrameCount() {
        return ((GifDrawable)drawable).getFrameCount();
    }
}

File: example_kotlin_swift/android/app/src/main/kotlin/com/alibaba/power_image_example_kotlin_swift/example_kotlin_swift/GlideMultiFrameImage.kt
package com.taobao.power_image_example

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.drawable.Drawable
import com.bumptech.glide.load.resource.gif.GifDrawable
import com.taobao.power_image.loader.FlutterMultiFrameImage
import java.lang.reflect.Field
import java.lang.reflect.Method

/**
 */
class GlideMultiFrameImage(drawable: GifDrawable?, needRecycle: Boolean) : FlutterMultiFrameImage(drawable, needRecycle) {
    companion object {
        private var sGifState: Field? = null
        private var sCurrentFrameMethod: Method? = null
        private var sFrameloader: Field? = null

        init {
            try {
                sGifState = GifDrawable::class.java.getDeclaredField("state")
                sGifState?.isAccessible = true
                sFrameloader = sGifState?.type?.getDeclaredField("frameLoader")
                sFrameloader?.isAccessible = true
                sCurrentFrameMethod = sFrameloader?.type?.getDeclaredMethod("getCurrentFrame")
                sCurrentFrameMethod?.isAccessible = true
            } catch (e1: Throwable) {
                e1.printStackTrace()
            }
        }
    }

    private val tmpCanvas = Canvas()
    override fun getCurrentFrame(who: Drawable): Bitmap? {
        who.draw(tmpCanvas)
        var bitmap: Bitmap? = null
        try {
            val gifState = sGifState!![who]
            val frameLoader = sFrameloader!![gifState]
            bitmap = sCurrentFrameMethod!!.invoke(frameLoader) as Bitmap
        } catch (t: Throwable) {
            t.printStackTrace()
        }
        return bitmap
    }

    protected override fun onStart(who: Drawable) {
        (who as GifDrawable).start()
    }

    protected override fun onRelease(who: Drawable) {
        (who as GifDrawable).stop()
    }

    override fun getFrameCount(): Int {
        return (drawable as GifDrawable).frameCount
    }
}